dali/image/bmp.cc

```diff
@@ -14,29 +14,109 @@
+                       size_t palette_entry_size = 0) {
+  if (compression_type == BMP_COMPRESSION_RGB || compression_type == BMP_COMPRESSION_RLE8) {
+    if (bpp <= 8 && ncolors <= static_cast<size_t>((1<<bpp))) {
+      return is_color_palette(palette_start, ncolors, palette_entry_size) ? 3 : 1;
+    } else if (bpp == 24) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  } else if (compression_type == BMP_COMPRESSION_BITFIELDS) {
+    if (bpp == 16) {
+      return 3;
+    } else if (bpp == 32) {
+      return 4;
+    }
+  }
+
+  DALI_WARN(make_string(
+    "configuration not supported. bpp:", bpp, "compression_type:", compression_type,
+    "ncolors:", ncolors));
+  return 0;
+}
+
+}  // namespace
+
 // https://en.wikipedia.org/wiki/BMP_file_format#DIB_header_(bitmap_information_header)
 
 BmpImage::BmpImage(const uint8_t *encoded_buffer, size_t length, DALIImageType image_type)
   : GenericImage(encoded_buffer, length, image_type) {}
 
 Image::Shape BmpImage::PeekShapeImpl(const uint8_t *bmp, size_t length) const {
   DALI_ENFORCE(bmp != nullptr);
-
-  uint32_t header_size = ReadValueLE<uint32_t>(bmp + 14);
-  int64_t h = 0, w = 0;
-  // BITMAPCOREHEADER: | 32u header | 16u width | 16u height | ...
+  auto ptr = bmp + 14;
+  uint32_t header_size = ConsumeValue<uint32_t>(ptr);
+  int64_t h = 0, w = 0, c = 0;
+  int bpp = 0, compression_type = BMP_COMPRESSION_RGB;
+  const uint8_t* palette_start = nullptr;
+  size_t ncolors = 0;
+  size_t palette_entry_size = 0;
   if (length >= 22 && header_size == 12) {
-    w = ReadValueLE<uint16_t>(bmp + 18);
-    h = ReadValueLE<uint16_t>(bmp + 20);
-    // BITMAPINFOHEADER and later: | 32u header | 32s width | 32s height | ...
+    // BITMAPCOREHEADER:
+    // | 32u header | 16u width | 16u height | 16u number of color planes | 16u bits per pixel
+    w = ConsumeValue<uint16_t>(ptr);
+    h = ConsumeValue<uint16_t>(ptr);
+    ptr += 2; // skip
+    bpp = ConsumeValue<uint16_t>(ptr);
+    if (bpp <= 8) {
+      palette_start = ptr;
+      palette_entry_size = 3;
+      ncolors = (1<<bpp);
+    }
   } else if (length >= 26 && header_size >= 40) {
-    w = ReadValueLE<int32_t>(bmp + 18);
-    h = abs(ReadValueLE<int32_t>(bmp + 22));
+    // BITMAPINFOHEADER and later:
+    // | 32u header | 32s width | 32s height | 16u number of color planes | 16u bits per pixel
+    // | 32u compression type
+    w = ConsumeValue<int32_t>(ptr);
+    h = abs(ConsumeValue<int32_t>(ptr));
+    ptr += 2;  // skip
+    bpp = ConsumeValue<uint16_t>(ptr);
+    compression_type = ConsumeValue<uint32_t>(ptr);
+    ptr += 12;  // skip
+    ncolors = ConsumeValue<uint32_t>(ptr);
+    ptr += header_size - 36;  // skip
+    if (bpp <= 8) {
+      palette_start = ptr;
+      palette_entry_size = 4;
+      ncolors = ncolors == 0 ? (1<<bpp) : ncolors;
+    }
+    c = number_of_channels(bpp, compression_type, palette_start, ncolors, palette_entry_size);
```

Comment:
I wonder how much sanity checks we need to do, like:
```
(ncolors * palette_entry_size) < length
```

Reply:
I have ` ncolors <= static_cast<size_t>((1<<bpp))) {` in line 51

Reply:
Added additional sanity check